File: scripts/gsd_ledger.js
#!/usr/bin/env node
/**
 * Author: RawNuke
 * Copyright (c) 2026 RawNuke. All rights reserved.
 *
 * Append rows to GSD-LEDGER.md.
 * Usage: node scripts/gsd_ledger.js --pr <n> --contributor <handle> --issue <n> --amount <n> --kind <kind> [--proposer <handle>] [--denomination GSD|USD] [--notes <text>]
 */

const fs = require("node:fs");
const path = require("node:path");

const LEDGER_PATH = path.resolve(__dirname, "..", "GSD-LEDGER.md");

const VALID_KINDS = new Set([
  "bounty",
  "proposal",
  "proposal-shipped",
  "implementation",
]);

function parseArgs(argv) {
  const args = {};
  for (let i = 2; i < argv.length; i++) {
    const key = argv[i];
    if (!key.startsWith("--")) continue;
    const val = argv[i + 1];
    if (val === undefined || val.startsWith("--")) {
      args[key.slice(2)] = true;
    } else {
      args[key.slice(2)] = val;
      i++;
    }
  }
  return args;
}

function fail(msg) {
  process.stderr.write(`Error: ${msg}\n`);
  process.exit(1);
}

function readLedger() {
  return fs.readFileSync(LEDGER_PATH, "utf-8");
}

function writeLedger(content) {
  fs.writeFileSync(LEDGER_PATH, content, "utf-8");
}

function today() {
  const d = new Date();
  const yyyy = d.getFullYear();
  const mm = String(d.getMonth() + 1).padStart(2, "0");
  const dd = String(d.getDate()).padStart(2, "0");
  return `${yyyy}-${mm}-${dd}`;
}

function parseExistingRows(lines) {
  const rows = [];
  let inTable = false;
  let headerFound = false;
  let separatorFound = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const trimmed = line.trim();

    if (!headerFound && trimmed.startsWith("| # |")) {
      headerFound = true;
      continue;
    }
    if (headerFound && !separatorFound && trimmed.startsWith("|---")) {
      separatorFound = true;
      inTable = true;
      continue;
    }
    if (!inTable) continue;
    if (!trimmed.startsWith("|")) break;

    const cells = line.split("|").map((c) => c.trim()).filter(Boolean);
    if (cells.length === 0) break;

    const num = cells[0];
    if (num.startsWith("_(") || num === "" || isNaN(parseInt(num, 10))) {
      continue;
    }

    rows.push({
      num: parseInt(num, 10),
      date: cells[1] || "",
      contributor: cells[2] || "",
      kind: cells[3] || "",
      pr: cells[4] || "",
      issue: cells[5] || "",
      amount: cells[6] || "",
      denomination: cells[7] || "",
      notes: cells[8] || "",
    });
  }

  return rows;
}

function buildRow(num, date, contributor, kind, pr, issue, amount, denomination, notes) {
  return `| ${num} | ${date} | ${contributor} | \`${kind}\` | ${pr} | ${issue} | ${amount} | ${denomination} | ${notes} |`;
}

function appendRows(lines, newRows) {
  let headerFound = false;
  let separatorFound = false;
  let insertAt = -1;
  let placeholderIdx = -1;

  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i].trim();

    if (!headerFound && trimmed.startsWith("| # |")) {
      headerFound = true;
      continue;
    }
    if (headerFound && !separatorFound && trimmed.startsWith("|---")) {
      separatorFound = true;
      insertAt = i + 1;
      continue;
    }
    if (separatorFound && trimmed.startsWith("|")) {
      if (trimmed.startsWith("| _(") || /^\|\s*_\(/.test(trimmed)) {
        placeholderIdx = i;
      }
      insertAt = i + 1;
    }
  }

  if (insertAt === -1) {
    fail("Could not find the ledger table in GSD-LEDGER.md");
  }

  if (placeholderIdx !== -1) {
    lines.splice(placeholderIdx, 1, ...newRows);
  } else {
    lines.splice(insertAt, 0, ...newRows);
  }

  return lines;
}

function main() {
  const args = parseArgs(process.argv);

  const pr = args.pr;
  const contributor = args.contributor;
  const issue = args.issue;
  const amount = args.amount;
  const kind = args.kind;
  const proposer = args.proposer;
  const denomination = args.denomination || "GSD";
  const notes = args.notes || "";

  if (!pr) fail("--pr is required");
  if (!contributor) fail("--contributor is required");
  if (!issue) fail("--issue is required");
  if (!amount) fail("--amount is required");
  if (!kind) fail("--kind is required");

  if (!VALID_KINDS.has(kind)) {
    fail(`--kind must be one of: ${[...VALID_KINDS].join(", ")}`);
  }

  if (denomination !== "GSD" && denomination !== "USD") {
    fail("--denomination must be GSD or USD");
  }

  const content = readLedger();
  const lines = content.split("\n");
  while (lines.length > 0 && lines[lines.length - 1] === "") {
    lines.pop();
  }
  const existingRows = parseExistingRows(lines);
  const lastNum = existingRows.length > 0
    ? Math.max(...existingRows.map((r) => r.num))
    : 0;

  const date = today();
  const newRows = [];

  newRows.push(
    buildRow(lastNum + 1, date, contributor, kind, pr, issue, amount, denomination, notes),
  );

  if (proposer) {
    newRows.push(
      buildRow(
        lastNum + 2,
        date,
        proposer,
        "proposal-shipped",
        pr,
        issue,
        "2",
        "GSD",
        `Proposer of PR #${pr}`,
      ),
    );
  }

  const updated = appendRows(lines, newRows);
  writeLedger(updated.join("\n") + "\n");

  process.stdout.write(
    `Appended ${newRows.length} row(s) to GSD-LEDGER.md\n`,
  );
}

main();

File: scripts/gsd_ledger.py
#!/usr/bin/env python3
"""Append a row to the GSD Ledger.

Usage:
    uv run scripts/gsd_ledger.py --pr 42 --contributor @alice --issue 7 --amount 1 --kind bounty
    uv run scripts/gsd_ledger.py --pr 42 --contributor @alice --issue 7 --amount 1 --kind bounty --denomination USD --notes "First ever GSD"
    uv run scripts/gsd_ledger.py --pr 42 --contributor @alice --issue 7 --amount 2 --kind implementation --proposer @bob --proposer-amount 2

The script reads GSD-LEDGER.md, finds the table, appends a new row with the
next sequential number, and writes the file back. It is append-only: existing
rows are never modified or deleted.
"""

import argparse
import re
import sys
from datetime import datetime, timezone
from pathlib import Path

LEDGER_PATH = Path(__file__).resolve().parent.parent / "GSD-LEDGER.md"

VALID_KINDS = {"bounty", "proposal", "proposal-shipped", "implementation"}

TABLE_HEADER_RE = re.compile(
    r"^\| # \| Date \| Contributor \| Kind \| PR \| Issue \| Amount \| Denomination \| Notes \|",
    re.MULTILINE,
)

TABLE_SEPARATOR_RE = re.compile(r"^\|[-\s|]+\|$", re.MULTILINE)


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Append a row to the GSD Ledger.")
    p.add_argument("--pr", type=int, required=True, help="Pull request number")
    p.add_argument("--contributor", required=True, help="Contributor handle (e.g. @alice)")
    p.add_argument("--issue", type=int, required=True, help="Issue number")
    p.add_argument("--amount", required=True, help="Amount awarded (e.g. 1, 2, 0.5)")
    p.add_argument("--kind", required=True, choices=sorted(VALID_KINDS), help="Award kind")
    p.add_argument("--denomination", default="GSD", help="Denomination (default: GSD)")
    p.add_argument("--notes", default="", help="Optional notes")
    p.add_argument(
        "--proposer",
        default=None,
        help="If set, also write a proposal-shipped row for this proposer",
    )
    p.add_argument(
        "--proposer-amount",
        default="2",
        help="Amount for the proposer row (default: 2)",
    )
    p.add_argument(
        "--ledger-path",
        type=Path,
        default=LEDGER_PATH,
        help="Path to GSD-LEDGER.md (default: auto-detect)",
    )
    return p.parse_args()


def get_next_row_number(content: str) -> int:
    """Find the highest row number in the table and return the next one."""
    numbers = re.findall(r"^\|\s*(\d+)\s*\|", content, re.MULTILINE)
    if not numbers:
        return 1
    return max(int(n) for n in numbers) + 1


def format_row(
    num: int,
    contributor: str,
    kind: str,
    pr: int,
    issue: int,
    amount: str,
    denomination: str,
    notes: str,
) -> str:
    """Format a single ledger row as a markdown table line."""
    date_str = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    if not contributor.startswith("@"):
        contributor = f"@{contributor}"
    notes = notes.replace("|", "\\|")
    pr_link = f"[#{pr}](https://github.com/jflournoy/for-funsies/pull/{pr})"
    issue_link = f"[#{issue}](https://github.com/jflournoy/for-funsies/issues/{issue})"
    return (
        f"| {num} | {date_str} | {contributor} | {kind} | {pr_link} | "
        f"{issue_link} | {amount} | {denomination} | {notes} |"
    )


def append_row(content: str, row: str) -> str:
    """Append a row to the table in the markdown content.

    The table ends at the last row that matches the table format.
    We insert the new row right after the last table row.
    """
    lines = content.split("\n")
    table_start = None
    table_end = None

    for i, line in enumerate(lines):
        if TABLE_HEADER_RE.match(line):
            table_start = i
        if table_start is not None and line.startswith("|"):
            table_end = i

    if table_start is None or table_end is None:
        print("ERROR: Could not find the ledger table in GSD-LEDGER.md", file=sys.stderr)
        sys.exit(1)

    # Find the separator line (it's right after the header)
    # and skip any "empty" placeholder rows
    insert_after = table_end

    # Skip placeholder rows like "| _(none yet...)_ |"
    for i in range(table_start + 2, table_end + 1):
        if "_(none yet" in lines[i] or lines[i].strip() == "":
            insert_after = i - 1
            break
    else:
        insert_after = table_end

    # Insert the new row after the last real row
    lines.insert(insert_after + 1, row)
    return "\n".join(lines)


def main() -> None:
    args = parse_args()

    ledger = args.ledger_path
    if not ledger.exists():
        print(f"ERROR: Ledger file not found: {ledger}", file=sys.stderr)
        sys.exit(1)

    content = ledger.read_text(encoding="utf-8")

    # Determine the starting row number
    next_num = get_next_row_number(content)

    # Build the primary row
    primary_row = format_row(
        num=next_num,
        contributor=args.contributor,
        kind=args.kind,
        pr=args.pr,
        issue=args.issue,
        amount=args.amount,
        denomination=args.denomination,
        notes=args.notes,
    )

    content = append_row(content, primary_row)
    next_num += 1

    # If a proposer is specified, also write a proposal-shipped row
    if args.proposer:
        proposer_row = format_row(
            num=next_num,
            contributor=args.proposer,
            kind="proposal-shipped",
            pr=args.pr,
            issue=args.issue,
            amount=args.proposer_amount,
            denomination=args.denomination,
            notes=f"Proposer of #{args.issue}",
        )
        content = append_row(content, proposer_row)

    ledger.write_text(content, encoding="utf-8")
    print(f"Appended row to {ledger}")
    if args.proposer:
        print(f"Appended proposer row for {args.proposer}")


if __name__ == "__main__":
    main()
